cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/autoinstall/HubPackage.java

```diff
@@ -0,0 +1,254 @@
+import io.cdap.cdap.internal.capability.HttpClients;
+import io.cdap.cdap.internal.guava.reflect.TypeToken;
+import io.cdap.cdap.proto.artifact.ArtifactRanges;
+import io.cdap.cdap.proto.id.NamespaceId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Represents a package json in the hub.
+ */
+public class HubPackage {
+  private static final Logger LOG = LoggerFactory.getLogger(HubPackage.class);
+  private static final Gson GSON = new Gson();
+
+  private final String name;
+  private final String version;
+  private final String description;
+  private final String label;
+  private final String author;
+  private final String org;
+  private final String cdapVersion;
+  private final long created;
+  private final boolean beta;
+  private final List<String> categories;
+  private final boolean paid;
+  private final String paidLink;
+  private transient ArtifactVersionRange versionRange;
+  private transient ArtifactVersion artifactVersion;
+
+  public HubPackage(String name, String version, String description, String label, String author, String org,
+                    String cdapVersion, long created, boolean beta, List<String> categories,
+                    boolean paid, String paidLink) throws InvalidArtifactRangeException {
+    this.name = name;
+    this.version = version;
+    this.artifactVersion = new ArtifactVersion(version);
+    this.description = description;
+    this.label = label;
+    this.author = author;
+    this.org = org;
+    this.cdapVersion = cdapVersion;
+    this.versionRange = ArtifactVersionRange.parse(cdapVersion);
+    this.created = created;
+    this.beta = beta;
+    this.categories = categories;
+    this.paid = paid;
+    this.paidLink = paidLink;
+  }
+
+  private String getCdapVersion() {
+    return cdapVersion;
+  }
+
+  public ArtifactVersion getArtifactVersion() {
+    if (artifactVersion == null) {
+      // artifactVersion will be null if the class was deserialized from json.
+      artifactVersion = new ArtifactVersion(version);
+
+    }
+    return artifactVersion;
+  }
+
+  public boolean versionIsInRange(String version) {
+    if (versionRange == null) {
+      // versionRange will be null if the class was deserialized from json.
+      if (cdapVersion == null) {
+        return false;
+      }
+      try {
+        versionRange = ArtifactVersionRange.parse(cdapVersion);
+      } catch (InvalidArtifactRangeException e) {
+        LOG.error(String.format("%s has invalid artifact range", name), e);
+        return false;
+      }
+    }
+    return versionRange.versionIsInRange(new ArtifactVersion(version));
+  }
+  public void installPlugin(URL url, ArtifactRepository artifactRepository, File tmpDir) throws Exception {
```

Comment:
Please add javadoc for non trivial public methods (e.g. not a simple getter). Also, add an empty to between consecutive methods.

Reply:
javadoc added